src/main/scala/firrtl/passes/memlib/ResolveMemoryReference.scala

```diff
@@ -21,28 +21,40 @@ class ResolveMemoryReference extends Transform {
 
   type AnnotatedMemories = collection.mutable.ArrayBuffer[(String, DefAnnotatedMemory)]
 
+  private def found(map: Map[String, Set[String]], key: String, value: String): Boolean =
+    map.get(key).map(_.contains(value)).getOrElse(false)
+
   /** If a candidate memory is identical except for name to another, add an
     *   annotation that references the name of the other memory.
     */
-  def updateMemStmts(mname: String, uniqueMems: AnnotatedMemories, noDeDupeMems: Seq[String])(s: Statement): Statement = s match {
+  def updateMemStmts(mname: String,
+                     uniqueMems: AnnotatedMemories,
+                     noDedupMap: Map[String, Set[String]])
+                    (s: Statement): Statement = s match {
     case m: DefAnnotatedMemory =>
-      uniqueMems find (x => eqMems(x._2, m, noDeDupeMems)) match {
+      uniqueMems.find { case (mname2, m2) =>
+        !found(noDedupMap, mname2, m2.name) &&
+        !found(noDedupMap, mname, m.name) &&
```

Comment:
The check on `m` could be hoisted out of the find right?
Not sure if that would be more or less clear, it took me a second to read it this way, so I think it might help readability slightly although it might be uglier scala.

Reply:
My previous response was actually just to Colin, you responded between my loading the page and submitting my comment 😛. In any case, this is definitely inefficient and originally I was just fixing the issue at hand but I'm wondering if just cleaning up the implementation wouldn't make more sense...